src/main/java/frc/team997/lib/trajectory/MotionProfile.java

```diff
@@ -0,0 +1,176 @@
+/**
+Copyright 2022 FRC Team 997
+
+This program is free software: 
+you can redistribute it and/or modify it under the terms of the 
+GNU General Public License as published by the Free Software Foundation, 
+either version 3 of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful, 
+but WITHOUT ANY WARRANTY; without even the implied warranty of 
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
+See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with SpartanLib2. 
+If not, see <https://www.gnu.org/licenses/>.
+*/
+package frc.team997.lib.trajectory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A motion profile provides a trajectory, most often used as the setpoint for a PID (or similar)
+ * controller.
+ *
+ * <p>This can be useful to avoid control effort saturation, or to decrease the initial large input
+ * from a PID's P term. Can also be useful to remain within physical constraints of a mechanism.
+ *
+ * <p>This class only contains constructors out of user-defined and user-calculated phases, but
+ * {@link AsymmetricTrapezoidProfile} and {@link TrapezoidProfile} extend this to create profiles
+ * out of constraints and desired states.
+ */
+public class MotionProfile {
+    protected List<ProfilePhase> phases;
+    private State initialState;
+
+    /** Holds position and velocity outputs from a MotionProfile. */
+    public static class State {
+        public final double position;
+        public final double velocity;
+
+        /**
+         * Constructs a State out of a given position and velocity.
+         *
+         * @param position Position of the profile.
+         * @param velocity Velocity of the profile.
+         */
+        public State(double position, double velocity) {
+            this.position = position;
+            this.velocity = velocity;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            double epsilon = 0.0001;
+            if (other instanceof State) {
+                State rhs = (State) other;
+                return Math.abs(this.position - rhs.position) < epsilon
+                        && Math.abs(this.velocity - rhs.velocity) < epsilon;
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(position, velocity);
+        }
+
+        @Override
+        public String toString() {
+            return "State[position: " + position + ", velocity: " + velocity + ", time: ";
+        }
+    }
+
+    /**
+     * Constructor for a MotionProfile.
+     *
+     * @param initialState The initial state of the motion profile.
+     * @param phases The ProfilePhases to start the profile with.
+     */
+    public MotionProfile(State initialState, ProfilePhase... phases) {
```

Comment:
Constructor for a MotionProfile with arbitrary phases

Reply:
also, ProfilePhase is effectively final (when not subclassed), so it's not to start the profile with but it's the phases for the profile